File: mobile/lib/core/activity_feed_utils.dart
import 'package:flutter/material.dart';
import 'package:hsdash_mobile/config/theme.dart';
import 'package:hsdash_mobile/core/calendar_utils.dart';
import 'package:hsdash_mobile/models/task.dart';
import 'package:hsdash_mobile/models/task_activity.dart';
import 'package:hsdash_mobile/models/tasks_query.dart';
import 'package:hsdash_mobile/models/team_member.dart';

enum ActivityPeriodFilter { today, week, month, all }

enum ActivityTypeFilter { all, assigned, started, completed, delayed }

enum ActivityViewMode { team, event, timeline }

enum OpsActivityKind { assigned, started, completed, delayed }

enum MemberHealthStatus { available, busy, delayed, noActivity }

class OpsActivityEntry {
  const OpsActivityEntry({
    required this.id,
    required this.taskId,
    required this.eventId,
    required this.kind,
    required this.timestamp,
    required this.memberId,
    required this.memberName,
    this.memberTeam,
    this.eventName,
    required this.taskName,
    this.synthetic = false,
  });

  final String id;
  final String taskId;
  final String eventId;
  final OpsActivityKind kind;
  final DateTime timestamp;
  final String memberId;
  final String memberName;
  final String? memberTeam;
  final String? eventName;
  final String taskName;
  final bool synthetic;

  factory OpsActivityEntry.fromActivity(TaskActivity a) {
    return OpsActivityEntry(
      id: a.id,
      taskId: a.taskId,
      eventId: a.eventId ?? '',
      kind: classifyActivityKind(a),
      timestamp: a.createdAtLocal ?? DateTime.fromMillisecondsSinceEpoch(0),
      memberId: a.actorId ?? a.assigneeId ?? 'unknown',
      memberName: a.actorName ?? 'Unknown',
      memberTeam: a.actorTeam ?? a.assignedTeam,
      eventName: a.clientName,
      taskName: a.taskLabel,
    );
  }

  factory OpsActivityEntry.fromTaskAssignment(Task task) {
    final created = task.createdAtLocal ?? DateTime.fromMillisecondsSinceEpoch(0);
    return OpsActivityEntry(
      id: 'assign-${task.id}',
      taskId: task.id,
      eventId: task.eventId,
      kind: OpsActivityKind.assigned,
      timestamp: created,
      memberId: task.assignedToId ?? 'unassigned',
      memberName: task.assigneeName ?? 'Unassigned',
      memberTeam: task.assignedTeam,
      eventName: task.clientName,
      taskName: task.label,
      synthetic: true,
    );
  }
}

OpsActivityKind classifyActivityKind(TaskActivity a) {
  if (a.newStatus == 'DELAYED') return OpsActivityKind.delayed;
  if (a.isCompleteEvent) return OpsActivityKind.completed;
  if (a.isStartEvent) return OpsActivityKind.started;
  if (a.isLateStart) return OpsActivityKind.delayed;
  return OpsActivityKind.started;
}

Color opsKindColor(OpsActivityKind kind) {
  switch (kind) {
    case OpsActivityKind.assigned:
      return AppColors.textMuted;
    case OpsActivityKind.started:
      return const Color(0xFF2563EB);
    case OpsActivityKind.completed:
      return AppColors.emerald;
    case OpsActivityKind.delayed:
      return AppColors.rose;
  }
}

String opsKindLabel(OpsActivityKind kind) {
  switch (kind) {
    case OpsActivityKind.assigned:
      return 'Assigned';
    case OpsActivityKind.started:
      return 'Started';
    case OpsActivityKind.completed:
      return 'Completed';
    case OpsActivityKind.delayed:
      return 'Delayed';
  }
}

class OpsDashboardFilters {
  const OpsDashboardFilters({
    this.period = ActivityPeriodFilter.today,
    this.eventId,
    this.memberId,
    this.type = ActivityTypeFilter.all,
    this.search = '',
    this.excludeMemberId,
  });

  final ActivityPeriodFilter period;
  final String? eventId;
  final String? memberId;
  final ActivityTypeFilter type;
  final String search;
  /// Hide this user's own actions (coordinator viewing team pulse).
  final String? excludeMemberId;

  OpsDashboardFilters copyWith({
    ActivityPeriodFilter? period,
    String? eventId,
    String? memberId,
    ActivityTypeFilter? type,
    String? search,
    String? excludeMemberId,
    bool clearEvent = false,
    bool clearMember = false,
  }) {
    return OpsDashboardFilters(
      period: period ?? this.period,
      eventId: clearEvent ? null : (eventId ?? this.eventId),
      memberId: clearMember ? null : (memberId ?? this.memberId),
      type: type ?? this.type,
      search: search ?? this.search,
      excludeMemberId: excludeMemberId ?? this.excludeMemberId,
    );
  }
}

class OpsOverviewMetrics {
  const OpsOverviewMetrics({
    required this.activeMembers,
    required this.assignedToday,
    required this.startedToday,
    required this.completedToday,
    required this.delayedTasks,
    required this.idleMembers,
  });

  final int activeMembers;
  final int assignedToday;
  final int startedToday;
  final int completedToday;
  final int delayedTasks;
  final int idleMembers;
}

class TeamHealthMetrics {
  const TeamHealthMetrics({
    required this.available,
    required this.busy,
    required this.delayed,
    required this.noActivity,
  });

  final int available;
  final int busy;
  final int delayed;
  final int noActivity;
}

class MemberOpsGroup {
  const MemberOpsGroup({
    required this.memberId,
    required this.memberName,
    required this.roleLabel,
    required this.openTasks,
    required this.startedInPeriod,
    required this.completedInPeriod,
    required this.lastActivity,
    required this.entries,
    required this.health,
    required this.taskTimelines,
  });

  final String memberId;
  final String memberName;
  final String roleLabel;
  final int openTasks;
  final int startedInPeriod;
  final int completedInPeriod;
  final DateTime? lastActivity;
  final List<OpsActivityEntry> entries;
  final MemberHealthStatus health;
  final List<TaskTimeline> taskTimelines;
}

class TaskTimeline {
  const TaskTimeline({
    required this.taskId,
    required this.eventName,
    required this.taskName,
    required this.steps,
  });

  final String taskId;
  final String eventName;
  final String taskName;
  final List<OpsActivityEntry> steps;
}

class EventOpsGroup {
  const EventOpsGroup({
    required this.eventId,
    required this.eventName,
    required this.assignedMembers,
    required this.startedMembers,
    required this.completedMembers,
    required this.delayedMembers,
    required this.entries,
  });

  final String eventId;
  final String eventName;
  final List<String> assignedMembers;
  final List<String> startedMembers;
  final List<String> completedMembers;
  final List<String> delayedMembers;
  final List<OpsActivityEntry> entries;
}

class OpsDashboardData {
  const OpsDashboardData({
    required this.overview,
    required this.health,
    required this.members,
    required this.events,
    required this.timeline,
    required this.eventOptions,
    required this.memberOptions,
  });

  final OpsOverviewMetrics overview;
  final TeamHealthMetrics health;
  final List<MemberOpsGroup> members;
  final List<EventOpsGroup> events;
  final List<OpsActivityEntry> timeline;
  final List<({String id, String label})> eventOptions;
  final List<({String id, String label})> memberOptions;
}

bool _inPeriod(DateTime at, ActivityPeriodFilter period) {
  final now = DateTime.now();
  switch (period) {
    case ActivityPeriodFilter.today:
      return localDayKey(at) == localDayKey(now);
    case ActivityPeriodFilter.week:
      return at.isAfter(now.subtract(const Duration(days: 7)));
    case ActivityPeriodFilter.month:
      return at.isAfter(now.subtract(const Duration(days: 30)));
    case ActivityPeriodFilter.all:
      return true;
  }
}

bool _isToday(DateTime at) => localDayKey(at) == localDayKey(DateTime.now());

String teamLabel(String? teamKey) {
  if (teamKey == null) return 'Team';
  return TaskTeam.labels[teamKey] ?? teamKey.replaceAll('_TEAM', '').replaceAll('_', ' ');
}

String memberRoleLabel(TeamMember? roster, String? teamKey) {
  if (roster?.designation != null && roster!.designation!.isNotEmpty) {
    return roster.designation!;
  }
  return teamLabel(teamKey);
}

List<OpsActivityEntry> buildOpsEntries({
  required List<TaskActivity> activities,
  required List<Task> tasks,
}) {
  final activityTaskIds = activities.map((a) => a.taskId).toSet();
  final entries = activities.map(OpsActivityEntry.fromActivity).toList();

  for (final task in tasks) {
    if (task.assignedToId == null) continue;
    if (activityTaskIds.contains(task.id)) continue;
    if (task.createdAtLocal == null) continue;
    entries.add(OpsActivityEntry.fromTaskAssignment(task));
  }

  entries.sort((a, b) => b.timestamp.compareTo(a.timestamp));
  return entries;
}

List<OpsActivityEntry> applyOpsFilters(
  List<OpsActivityEntry> entries,
  OpsDashboardFilters filters,
) {
  final q = filters.search.trim().toLowerCase();
  return entries.where((e) {
    if (!_inPeriod(e.timestamp, filters.period)) return false;
    if (filters.excludeMemberId != null && e.memberId == filters.excludeMemberId) return false;
    if (filters.eventId != null && e.eventId != filters.eventId) return false;
    if (filters.memberId != null && e.memberId != filters.memberId) return false;
    if (filters.type != ActivityTypeFilter.all) {
      final match = switch (filters.type) {
        ActivityTypeFilter.assigned => e.kind == OpsActivityKind.assigned,
        ActivityTypeFilter.started => e.kind == OpsActivityKind.started,
        ActivityTypeFilter.completed => e.kind == OpsActivityKind.completed,
        ActivityTypeFilter.delayed => e.kind == OpsActivityKind.delayed,
        ActivityTypeFilter.all => true,
      };
      if (!match) return false;
    }
    if (q.isNotEmpty) {
      final hay = '${e.memberName} ${e.eventName ?? ''} ${e.taskName}'.toLowerCase();
      if (!hay.contains(q)) return false;
    }
    return true;
  }).toList();
}

MemberHealthStatus _memberHealth({
  required List<Task> memberTasks,
  required List<OpsActivityEntry> memberEntries,
  required ActivityPeriodFilter period,
}) {
  final open = memberTasks.where((t) => t.status != 'COMPLETED').toList();
  if (open.isEmpty) return MemberHealthStatus.available;
  if (open.any((t) => t.status == 'DELAYED')) return MemberHealthStatus.delayed;
  if (open.any((t) => t.status == 'IN_PROGRESS')) return MemberHealthStatus.busy;
  final hasActivity = memberEntries.any((e) => _inPeriod(e.timestamp, period));
  if (open.isNotEmpty && !hasActivity) return MemberHealthStatus.noActivity;
  return MemberHealthStatus.available;
}

List<TaskTimeline> _buildTaskTimelines(List<OpsActivityEntry> entries, List<Task> tasks) {
  final tasksById = {for (final t in tasks) t.id: t};
  final byTask = <String, List<OpsActivityEntry>>{};
  for (final e in entries) {
    byTask.putIfAbsent(e.taskId, () => []).add(e);
  }
  return byTask.entries.map((kv) {
    final steps = [...kv.value]..sort((a, b) => a.timestamp.compareTo(b.timestamp));
    final task = tasksById[kv.key];
    if (task != null && task.createdAtLocal != null && !steps.any((s) => s.kind == OpsActivityKind.assigned)) {
      steps.insert(0, OpsActivityEntry.fromTaskAssignment(task));
    }
    final first = steps.first;
    return TaskTimeline(
      taskId: kv.key,
      eventName: first.eventName ?? 'Event',
      taskName: first.taskName,
      steps: steps,
    );
  }).toList()
    ..sort((a, b) {
      final aLast = a.steps.last.timestamp;
      final bLast = b.steps.last.timestamp;
      return bLast.compareTo(aLast);
    });
}

OpsDashboardData buildOpsDashboard({
  required List<TaskActivity> activities,
  required List<Task> tasks,
  required OpsDashboardFilters filters,
  List<TeamMember> roster = const [],
}) {
  final allEntries = buildOpsEntries(activities: activities, tasks: tasks);
  final filtered = applyOpsFilters(allEntries, filters);

  final rosterById = {for (final m in roster) m.id: m};
  final memberIds = <String>{};
  for (final e in filtered) {
    if (e.memberId != 'unknown' && e.memberId != 'unassigned') memberIds.add(e.memberId);
  }
  for (final t in tasks) {
    if (t.assignedToId != null) memberIds.add(t.assignedToId!);
  }
  if (filters.excludeMemberId != null) {
    memberIds.remove(filters.excludeMemberId);
  }

  final eventOptionsMap = <String, String>{};
  for (final e in allEntries) {
    if (e.eventId.isNotEmpty) eventOptionsMap[e.eventId] = e.eventName ?? 'Event';
  }
  for (final t in tasks) {
    eventOptionsMap[t.eventId] = t.clientName ?? 'Event';
  }

  final memberOptions = memberIds.map((id) {
    String name = rosterById[id]?.name ?? 'Unknown';
    for (final e in allEntries) {
      if (e.memberId == id) {
        name = e.memberName;
        break;
      }
    }
    for (final t in tasks) {
      if (t.assignedToId == id && t.assigneeName != null) {
        name = t.assigneeName!;
        break;
      }
    }
    return (id: id, label: name);
  }).toList()
    ..sort((a, b) => a.label.compareTo(b.label));

  final eventOptions = eventOptionsMap.entries.map((e) => (id: e.key, label: e.value)).toList()
    ..sort((a, b) => a.label.compareTo(b.label));

  var assignedToday = 0;
  var startedToday = 0;
  var completedToday = 0;
  var delayedTasks = tasks.where((t) => t.status == 'DELAYED').length;
  var activeMembers = 0;
  var idleMembers = 0;

  final members = <MemberOpsGroup>[];
  for (final id in memberIds) {
    final memberTasks = tasks.where((t) => t.assignedToId == id).toList();
    final memberEntries = filtered.where((e) => e.memberId == id).toList();
    final openCount = memberTasks.where((t) => t.status != 'COMPLETED').length;
    final started = memberEntries.where((e) => e.kind == OpsActivityKind.started).length;
    final completed = memberEntries.where((e) => e.kind == OpsActivityKind.completed).length;
    final last = memberEntries.isEmpty ? null : memberEntries.first.timestamp;
    final teamKey = rosterById[id]?.team ?? memberEntries.firstOrNull?.memberTeam ?? memberTasks.firstOrNull?.assignedTeam;

    final health = _memberHealth(memberTasks: memberTasks, memberEntries: memberEntries, period: filters.period);
    if (memberEntries.isNotEmpty) activeMembers++;
    if (health == MemberHealthStatus.noActivity) idleMembers++;

    members.add(
      MemberOpsGroup(
        memberId: id,
        memberName: rosterById[id]?.name ?? memberEntries.firstOrNull?.memberName ?? 'Unknown',
        roleLabel: memberRoleLabel(rosterById[id], teamKey),
        openTasks: openCount,
        startedInPeriod: started,
        completedInPeriod: completed,
        lastActivity: last,
        entries: memberEntries,
        health: health,
        taskTimelines: _buildTaskTimelines(memberEntries, tasks),
      ),
    );
  }

  members.sort((a, b) {
    return (b.lastActivity ?? DateTime.fromMillisecondsSinceEpoch(0))
        .compareTo(a.lastActivity ?? DateTime.fromMillisecondsSinceEpoch(0));
  });

  for (final e in filtered) {
    if (e.kind == OpsActivityKind.assigned && _isToday(e.timestamp)) assignedToday++;
    if (e.kind == OpsActivityKind.started && _isToday(e.timestamp)) startedToday++;
    if (e.kind == OpsActivityKind.completed && _isToday(e.timestamp)) completedToday++;
  }

  var available = 0;
  var busy = 0;
  var delayed = 0;
  var noActivity = 0;
  for (final m in members) {
    switch (m.health) {
      case MemberHealthStatus.available:
        available++;
      case MemberHealthStatus.busy:
        busy++;
      case MemberHealthStatus.delayed:
        delayed++;
      case MemberHealthStatus.noActivity:
        noActivity++;
    }
  }

  final eventsMap = <String, EventOpsGroup>{};
  for (final e in filtered) {
    if (e.eventId.isEmpty) continue;
    final existing = eventsMap[e.eventId];
    final assigned = {...?existing?.assignedMembers.toSet()};
    final started = {...?existing?.startedMembers.toSet()};
    final completed = {...?existing?.completedMembers.toSet()};
    final delayedMembers = {...?existing?.delayedMembers.toSet()};
    final entries = [...?existing?.entries, e];

    switch (e.kind) {
      case OpsActivityKind.assigned:
        assigned.add(e.memberName);
      case OpsActivityKind.started:
        started.add(e.memberName);
      case OpsActivityKind.completed:
        completed.add(e.memberName);
      case OpsActivityKind.delayed:
        delayedMembers.add(e.memberName);
    }

    eventsMap[e.eventId] = EventOpsGroup(
      eventId: e.eventId,
      eventName: e.eventName ?? existing?.eventName ?? 'Event',
      assignedMembers: assigned.toList()..sort(),
      startedMembers: started.toList()..sort(),
      completedMembers: completed.toList()..sort(),
      delayedMembers: delayedMembers.toList()..sort(),
      entries: entries,
    );
  }

  final events = eventsMap.values.toList()
    ..sort((a, b) => a.eventName.compareTo(b.eventName));

  return OpsDashboardData(
    overview: OpsOverviewMetrics(
      activeMembers: activeMembers,
      assignedToday: assignedToday,
      startedToday: startedToday,
      completedToday: completedToday,
      delayedTasks: delayedTasks,
      idleMembers: idleMembers,
    ),
    health: TeamHealthMetrics(
      available: available,
      busy: busy,
      delayed: delayed,
      noActivity: noActivity,
    ),
    members: members,
    events: events,
    timeline: filtered,
    eventOptions: eventOptions,
    memberOptions: memberOptions,
  );
}

extension _FirstOrNull<E> on Iterable<E> {
  E? get firstOrNull {
    final it = iterator;
    if (!it.moveNext()) return null;
    return it.current;
  }
}

File: frontend/src/lib/activityFeedUtils.ts
import { calendarDayKeyFromIso, localDayKey } from "@/lib/calendarUtils";
import type { AttendanceAlertRow } from "@/types/attendance";
import type { Task, TaskStatus, Team, User } from "@/types/domain";

export type ActivityPeriodFilter = "today" | "week" | "month" | "all";
export type ActivityTypeFilter = "all" | "assigned" | "started" | "completed" | "delayed" | "attendance";
export type OpsActivityKind = "assigned" | "started" | "completed" | "delayed" | "attendance";
export type MemberHealthStatus = "available" | "busy" | "delayed" | "noActivity";

export type TaskActivityRow = {
  id: string;
  taskId: string;
  previousStatus: TaskStatus | null;
  newStatus: TaskStatus;
  createdAt: string;
  actor: { id: string; name: string; email: string; team: Team | null } | null;
  task: {
    id: string;
    taskType: string;
    assignedTeam: string;
    assignedToId?: string | null;
    assignedTo?: { id: string; name: string } | null;
    eventId?: string;
    event: { id?: string; clientName: string; eventDate?: string } | null;
  };
};

export type OpsActivityEntry = {
  id: string;
  taskId: string;
  eventId: string;
  kind: OpsActivityKind;
  timestamp: Date;
  memberId: string;
  memberName: string;
  memberTeam?: string;
  eventName?: string;
  taskName: string;
  synthetic: boolean;
};

export type OpsDashboardFilters = {
  period: ActivityPeriodFilter;
  eventId?: string | null;
  memberId?: string | null;
  type: ActivityTypeFilter;
  search: string;
  excludeMemberId?: string | null;
};

export type OpsOverviewMetrics = {
  activeMembers: number;
  assignedToday: number;
  startedToday: number;
  completedToday: number;
  delayedTasks: number;
  idleMembers: number;
};

export type TeamHealthMetrics = {
  available: number;
  busy: number;
  delayed: number;
  noActivity: number;
};

export type MemberOpsGroup = {
  memberId: string;
  memberName: string;
  roleLabel: string;
  openTasks: number;
  startedInPeriod: number;
  completedInPeriod: number;
  lastActivity: Date | null;
  entries: OpsActivityEntry[];
  health: MemberHealthStatus;
};

export type EventOpsGroup = {
  eventId: string;
  eventName: string;
  assignedMembers: string[];
  startedMembers: string[];
  completedMembers: string[];
  delayedMembers: string[];
  entries: OpsActivityEntry[];
};

export type OpsDashboardData = {
  overview: OpsOverviewMetrics;
  health: TeamHealthMetrics;
  members: MemberOpsGroup[];
  events: EventOpsGroup[];
  timeline: OpsActivityEntry[];
  eventOptions: { id: string; label: string }[];
  memberOptions: { id: string; label: string }[];
};

function inPeriod(at: Date, period: ActivityPeriodFilter) {
  const now = new Date();
  switch (period) {
    case "today":
      return localDayKey(at.getFullYear(), at.getMonth(), at.getDate()) === localDayKey(now.getFullYear(), now.getMonth(), now.getDate());
    case "week":
      return at.getTime() > now.getTime() - 7 * 86400000;
    case "month":
      return at.getTime() > now.getTime() - 30 * 86400000;
    case "all":
      return true;
  }
}

function isToday(at: Date) {
  const now = new Date();
  return localDayKey(at.getFullYear(), at.getMonth(), at.getDate()) === localDayKey(now.getFullYear(), now.getMonth(), now.getDate());
}

export function classifyActivityKind(a: TaskActivityRow): OpsActivityKind {
  if (a.newStatus === "DELAYED") return "delayed";
  if (a.previousStatus && a.newStatus === "COMPLETED") return "completed";
  if (a.previousStatus === "PENDING" && a.newStatus === "IN_PROGRESS") return "started";
  if (a.newStatus === "IN_PROGRESS") return "started";
  if (a.newStatus === "COMPLETED") return "completed";
  return "started";
}

export function opsKindLabel(kind: OpsActivityKind) {
  switch (kind) {
    case "assigned":
      return "Assigned";
    case "started":
      return "Started";
    case "completed":
      return "Completed";
    case "delayed":
      return "Delayed";
    case "attendance":
      return "Attendance";
  }
}

export function opsKindColor(kind: OpsActivityKind) {
  switch (kind) {
    case "assigned":
      return "text-zinc-600 bg-zinc-100 border-zinc-200";
    case "started":
      return "text-blue-700 bg-blue-50 border-blue-200";
    case "completed":
      return "text-emerald-700 bg-emerald-50 border-emerald-200";
    case "delayed":
      return "text-rose-700 bg-rose-50 border-rose-200";
    case "attendance":
      return "text-amber-800 bg-amber-50 border-amber-200";
  }
}

export function healthLabel(status: MemberHealthStatus) {
  switch (status) {
    case "available":
      return "Available";
    case "busy":
      return "In progress";
    case "delayed":
      return "Delayed";
    case "noActivity":
      return "No recent activity";
  }
}

export function healthColor(status: MemberHealthStatus) {
  switch (status) {
    case "available":
      return "text-emerald-700 bg-emerald-50 border-emerald-200";
    case "busy":
      return "text-blue-700 bg-blue-50 border-blue-200";
    case "delayed":
      return "text-rose-700 bg-rose-50 border-rose-200";
    case "noActivity":
      return "text-amber-800 bg-amber-50 border-amber-200";
  }
}

function teamLabel(teamKey: string | null | undefined) {
  if (!teamKey) return "Team";
  return teamKey.replaceAll("_", " ");
}

function memberRoleLabel(roster: User | undefined, teamKey: string | null | undefined) {
  if (roster?.designation) return roster.designation;
  return teamLabel(teamKey);
}

export function opsActivityFromRow(a: TaskActivityRow): OpsActivityEntry {
  return {
    id: a.id,
    taskId: a.taskId,
    eventId: a.task.event?.id ?? a.task.eventId ?? "",
    kind: classifyActivityKind(a),
    timestamp: new Date(a.createdAt),
    memberId: a.actor?.id ?? a.task.assignedToId ?? a.task.assignedTo?.id ?? "unknown",
    memberName: a.actor?.name ?? a.task.assignedTo?.name ?? "Unknown",
    memberTeam: a.actor?.team ?? a.task.assignedTeam,
    eventName: a.task.event?.clientName,
    taskName: a.task.taskType.replaceAll("_", " "),
    synthetic: false,
  };
}

function opsActivityFromAttendance(alert: AttendanceAlertRow): OpsActivityEntry {
  return {
    id: alert.id,
    taskId: "",
    eventId: "",
    kind: "attendance",
    timestamp: new Date(alert.occurredAt),
    memberId: alert.userId,
    memberName: alert.user?.name ?? "Unknown",
    memberTeam: alert.user?.team ?? undefined,
    eventName: "Attendance",
    taskName: alert.message,
    synthetic: false,
  };
}

function opsActivityFromTask(task: Task): OpsActivityEntry | null {
  if (!task.assignedToId) return null;
  return {
    id: `assign-${task.id}`,
    taskId: task.id,
    eventId: task.eventId,
    kind: "assigned",
    timestamp: new Date(task.createdAt),
    memberId: task.assignedToId,
    memberName: task.assignedTo?.name ?? "Unassigned",
    memberTeam: task.assignedTeam,
    eventName: task.event?.clientName,
    taskName: task.taskType.replaceAll("_", " "),
    synthetic: true,
  };
}

export function buildOpsEntries(activities: TaskActivityRow[], tasks: Task[], attendanceAlerts: AttendanceAlertRow[] = []) {
  const activityTaskIds = new Set(activities.map((a) => a.taskId));
  const entries = activities.map(opsActivityFromRow);
  for (const task of tasks) {
    if (activityTaskIds.has(task.id)) continue;
    const synthetic = opsActivityFromTask(task);
    if (synthetic) entries.push(synthetic);
  }
  for (const alert of attendanceAlerts) {
    entries.push(opsActivityFromAttendance(alert));
  }
  entries.sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());
  return entries;
}

export function applyOpsFilters(entries: OpsActivityEntry[], filters: OpsDashboardFilters) {
  const q = filters.search.trim().toLowerCase();
  return entries.filter((e) => {
    if (!inPeriod(e.timestamp, filters.period)) return false;
    if (filters.excludeMemberId && e.memberId === filters.excludeMemberId) return false;
    if (filters.eventId && e.eventId !== filters.eventId) return false;
    if (filters.memberId && e.memberId !== filters.memberId) return false;
    if (filters.type !== "all" && e.kind !== filters.type) return false;
    if (q) {
      const hay = `${e.memberName} ${e.eventName ?? ""} ${e.taskName}`.toLowerCase();
      if (!hay.includes(q)) return false;
    }
    return true;
  });
}

function memberHealth(
  memberTasks: Task[],
  memberEntries: OpsActivityEntry[],
  period: ActivityPeriodFilter,
): MemberHealthStatus {
  const open = memberTasks.filter((t) => t.status !== "COMPLETED");
  if (!open.length) return "available";
  if (open.some((t) => t.status === "DELAYED")) return "delayed";
  if (open.some((t) => t.status === "IN_PROGRESS")) return "busy";
  const hasActivity = memberEntries.some((e) => inPeriod(e.timestamp, period));
  if (open.length && !hasActivity) return "noActivity";
  return "available";
}

export function buildOpsDashboard(
  activities: TaskActivityRow[],
  tasks: Task[],
  filters: OpsDashboardFilters,
  roster: User[] = [],
  attendanceAlerts: AttendanceAlertRow[] = [],
): OpsDashboardData {
  const allEntries = buildOpsEntries(activities, tasks, attendanceAlerts);
  const filtered = applyOpsFilters(allEntries, filters);
  const rosterById = new Map(roster.map((m) => [m.id, m]));
  const memberIds = new Set<string>();
  for (const e of filtered) {
    if (e.memberId !== "unknown" && e.memberId !== "unassigned") memberIds.add(e.memberId);
  }
  for (const t of tasks) {
    if (t.assignedToId) memberIds.add(t.assignedToId);
  }
  if (filters.excludeMemberId) memberIds.delete(filters.excludeMemberId);

  const eventOptionsMap = new Map<string, string>();
  for (const e of allEntries) {
    if (e.eventId) eventOptionsMap.set(e.eventId, e.eventName ?? "Event");
  }
  for (const t of tasks) {
    eventOptionsMap.set(t.eventId, t.event?.clientName ?? "Event");
  }

  const memberOptions = [...memberIds]
    .map((id) => {
      let name = rosterById.get(id)?.name ?? "Unknown";
      for (const e of allEntries) {
        if (e.memberId === id) {
          name = e.memberName;
          break;
        }
      }
      for (const t of tasks) {
        if (t.assignedToId === id && t.assignedTo?.name) {
          name = t.assignedTo.name;
          break;
        }
      }
      return { id, label: name };
    })
    .sort((a, b) => a.label.localeCompare(b.label));

  const eventOptions = [...eventOptionsMap.entries()]
    .map(([id, label]) => ({ id, label }))
    .sort((a, b) => a.label.localeCompare(b.label));

  let assignedToday = 0;
  let startedToday = 0;
  let completedToday = 0;
  const delayedTasks = tasks.filter((t) => t.status === "DELAYED").length;
  let activeMembers = 0;
  let idleMembers = 0;

  const members: MemberOpsGroup[] = [];
  for (const id of memberIds) {
    const memberTasks = tasks.filter((t) => t.assignedToId === id);
    const memberEntries = filtered.filter((e) => e.memberId === id);
    const openCount = memberTasks.filter((t) => t.status !== "COMPLETED").length;
    const started = memberEntries.filter((e) => e.kind === "started").length;
    const completed = memberEntries.filter((e) => e.kind === "completed").length;
    const last = memberEntries[0]?.timestamp ?? null;
    const teamKey =
      rosterById.get(id)?.team ?? memberEntries[0]?.memberTeam ?? memberTasks[0]?.assignedTeam ?? null;
    const health = memberHealth(memberTasks, memberEntries, filters.period);
    if (memberEntries.length) activeMembers++;
    if (health === "noActivity") idleMembers++;
    members.push({
      memberId: id,
      memberName: rosterById.get(id)?.name ?? memberEntries[0]?.memberName ?? "Unknown",
      roleLabel: memberRoleLabel(rosterById.get(id), teamKey),
      openTasks: openCount,
      startedInPeriod: started,
      completedInPeriod: completed,
      lastActivity: last,
      entries: memberEntries,
      health,
    });
  }

  members.sort(
    (a, b) => (b.lastActivity?.getTime() ?? 0) - (a.lastActivity?.getTime() ?? 0),
  );

  for (const e of filtered) {
    if (e.kind === "assigned" && isToday(e.timestamp)) assignedToday++;
    if (e.kind === "started" && isToday(e.timestamp)) startedToday++;
    if (e.kind === "completed" && isToday(e.timestamp)) completedToday++;
  }

  let available = 0;
  let busy = 0;
  let delayed = 0;
  let noActivity = 0;
  for (const m of members) {
    switch (m.health) {
      case "available":
        available++;
        break;
      case "busy":
        busy++;
        break;
      case "delayed":
        delayed++;
        break;
      case "noActivity":
        noActivity++;
        break;
    }
  }

  const eventsMap = new Map<string, EventOpsGroup>();
  for (const e of filtered) {
    if (!e.eventId) continue;
    const existing = eventsMap.get(e.eventId);
    const assigned = new Set(existing?.assignedMembers ?? []);
    const started = new Set(existing?.startedMembers ?? []);
    const completed = new Set(existing?.completedMembers ?? []);
    const delayedMembers = new Set(existing?.delayedMembers ?? []);
    const entries = [...(existing?.entries ?? []), e];
    switch (e.kind) {
      case "assigned":
        assigned.add(e.memberName);
        break;
      case "started":
        started.add(e.memberName);
        break;
      case "completed":
        completed.add(e.memberName);
        break;
      case "delayed":
        delayedMembers.add(e.memberName);
        break;
      case "attendance":
        break;
    }
    eventsMap.set(e.eventId, {
      eventId: e.eventId,
      eventName: e.eventName ?? existing?.eventName ?? "Event",
      assignedMembers: [...assigned].sort(),
      startedMembers: [...started].sort(),
      completedMembers: [...completed].sort(),
      delayedMembers: [...delayedMembers].sort(),
      entries,
    });
  }

  const events = [...eventsMap.values()].sort((a, b) => a.eventName.localeCompare(b.eventName));

  return {
    overview: {
      activeMembers,
      assignedToday,
      startedToday,
      completedToday,
      delayedTasks,
      idleMembers,
    },
    health: { available, busy, delayed, noActivity },
    members,
    events,
    timeline: filtered,
    eventOptions,
    memberOptions,
  };
}

export function formatActivityWhen(d: Date) {
  return d.toLocaleString(undefined, { dateStyle: "medium", timeStyle: "short" });
}

export function formatDayKey(isoOrKey: string) {
  const key = isoOrKey.includes("T") ? calendarDayKeyFromIso(isoOrKey) : isoOrKey;
  const [y, m, d] = key.split("-").map(Number);
  return new Date(y, m - 1, d).toLocaleDateString(undefined, {
    weekday: "short",
    month: "short",
    day: "numeric",
    year: "numeric",
  });
}
